utils/zerovec/src/varzerovec/owned.rs

```diff
@@ -115,6 +197,172 @@ impl<T: AsVarULE> VarZeroVecOwned<T> {
+
+            // The memory range of the affected index.
+            // When inserting: where the new index goes.
+            // When removing:  where the index being removed is.
+            // When replacing: unused.
+            let index_range = {
+                let index_start = slice_range.start.add(4 + 4 * index);
+                index_start..index_start.add(4)
+            };
+
+            unsafe fn shift_bytes(block: Range<*const u8>, to: *mut u8) {
+                debug_assert!(block.end >= block.start);
+                ptr::copy(block.start, to, block.end.offset_from(block.start) as usize);
+            }
+
+            if shift_type == ShiftType::Remove {
+                // Move the data before the element back by 4 to remove the index.
+                shift_bytes(index_range.end..prev_element_p.start, index_range.start);
+            }
+
+            // Shift data after the element to its new position.
+            shift_bytes(
+                prev_element_p.end..slice_range.end,
+                prev_element_p
+                    .start
+                    .offset((new_size as i64 + index_shift) as isize),
+            );
+
+            let first_affected_index = match shift_type {
+                ShiftType::Insert => {
+                    // Move data before the element forward by 4 to make space for a new index.
+                    shift_bytes(index_range.start..prev_element_p.start, index_range.end);
+
+                    *self.index_data_mut(index) = (prev_element.start as u32).into();
+                    self.set_len((len + 1) as u32);
+                    index + 1
+                }
+                ShiftType::Remove => {
+                    self.set_len((len - 1) as u32);
+                    index
+                }
+                ShiftType::Replace => index + 1,
+            };
+            // No raw pointer use should occur after this point (because of self.index_data and self.set_len).
+
+            // Set the new slice length. This must be done after shifting data around to avoid uninitialized data.
+            self.entire_slice.set_len(new_slice_len);
+
+            // Shift the affected indices.
+            self.shift_indices(first_affected_index, (shift - index_shift) as i32);
+        };
+
+        debug_assert!(self.verify_integrity());
+
+        // Return a mut slice to the new element data.
+        let element_pos = 4 + self.len() * 4 + self.element_position_unchecked(index);
+        &mut self.entire_slice[element_pos..element_pos + new_size as usize]
+    }
+
+    /// Checks the internal invariants of the vec to ensure safe code will not cause UB.
+    /// Returns whether integrity was verified.
+    ///
+    /// Note: an index is valid if it doesn't point to data past the end of the slice and is
+    /// less than or equal to all future indices. The length of the index segment is not part of each index.
+    fn verify_integrity(&self) -> bool {
```

Comment:
Debating if this is too overboard or if it should just depend on `SliceComponents::parse_byte_slice`